CoseSign1.Certificates/Local/Validators/X509ChainTrustValidator.cs

```diff
@@ -115,6 +112,23 @@ public class X509ChainTrustValidator : X509Certificate2MessageValidator
             return new CoseSign1ValidationResult(GetType(), true, "Certificate was Trusted.");
         }
 
+        // If we fail because chain build failed to reach the revocation server, retry in case the server is down.
+        if (ChainBuilder.ChainPolicy.RevocationMode != X509RevocationMode.NoCheck)
+        {
+            int maxAttempts = 3;
+            for (int i = 0;
+                i < maxAttempts && ChainBuilder.ChainStatus.Any(s => (s.Status & X509ChainStatusFlags.RevocationStatusUnknown) != 0);
+                i++)
+            {
+                if (ChainBuilder.Build(signingCertificate))
+                {
+                    return new CoseSign1ValidationResult(GetType(), true, "Certificate was Trusted.");
+                }
+
+                Thread.Sleep(1000);
+            }
+        }
+
         // Chain build failed, but if the only failure is Untrusted Root we may still pass.
         if (ChainBuilder.ChainStatus.All(st => st.Status.HasFlag(X509ChainStatusFlags.UntrustedRoot) || st.Status.HasFlag(X509ChainStatusFlags.NoError)))

```

Comment:
It looks like we are entering this block if we have X509ChainStatusFlags.UntrustedRoot as one of our flags, but it doesn't look like we are checking for any other flags. In the case where we fail to lookup CRL _and_ we have a user provided root, it looks like we may incorrectly pass validation.

Reply:
fixed